(DOCSP-19911): Use Realm React documentation (#1638)
mongodb/docs-realm: source/sdk/react-native/use-realm-hooks.txt

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 3
   :class: singlecol

Overview
--------
:npm:`Realm Hooks <package/@realm/react>` provides an easy-to-use API to perform
common Realm operations, such as querying or writing to a realm and listening to
realm objects.

Using Realm Hooks allows you to avoid creating boilerplate code, such as
creating your own listeners and state management. Realm React Hooks provides a
set of React Hooks, which return state-aware Realm data. This means that React
components using the Realm React Hooks will re-render whenever you change the
Realm data.

.. _react-native-setup-realm-hooks:

Setup Realm Hooks
-----------------

> To set up realm hooks, you can either start from scratch with a new application

Author: Yup, switched to Realm React & capitalized now. There should be no more instances of "Realm Hooks" but there's still some of hooks (just to describe the actual hook methods, not as the name of the library)

   tabs:
     - id: create-a-new-app-with-realm-hooks-preconfigured
       name: Create a New App with Realm Hooks
       content: |
         If you don't have an existing application, we recommend developing with
         the :github:`Realm Expo template <expo/examples/tree/master/with-realm>`.
         The Realm Expo template allows you to bootstrap your application with
         `Expo <https://docs.expo.dev/>`_ and Realm Hooks preconfigured.

         To initialize the Realm Expo template, read the :ref:`Quick Start
         with Expo <react-native-client-quick-start-expo>` documentation.

     - id: use-npm-to-add-realm-hooks-to-an-old-app
       name: Add Realm Hooks to an Existing App
       content: |
         To install Realm Hooks on an existing React Native application, run the
         following command in your terminal from the root of your application
         where your ``package.json`` file is located:

         .. code-block:: shell

            npm i @realm/react

.. _react-native-realm-context:

Create a Realm Configuration and Realm Context
----------------------------------------------

The ``createRealmContext()`` method creates a Realm :reactjs:`Context

> <docs/context.html>` object that opens a realm with a given :js-sdk:`Realm.Configuration

Author: * Got it, sorry I thought createRealmContext opened the realm itself. I've updated the wording now, let me know if that works. 
* I added an example of how to dynamically update the realm config via the `<RealmProvider />` component props in the [Using The Realm Provider](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-19911-realm-hooks-docs/sdk/react-native/use-realm-hooks/#using-the-realm-provider) section.

<Realm.html#~Configuration>`. The ``Realm Context`` object contains the following: 

- Various pre-built :reactjs:`Hooks <docs/hooks-intro.html>` that you can use to get access to the opened {+realm+}, query the {+realm+}, etc.

> - A :reactjs:`Context Provider <docs/context.html#contextprovider>` component that wraps around a component, and provides any of it's child components with access to the hooks.

Reviewer: ```suggestion
- A :reactjs:`Context Provider <docs/context.html#contextprovider>` component that wraps around a component, and provides any of its child components with access to the hooks.
```

Reviewer: Sorry for being a bit of a grammar nut, but `it's` is only used as a contraction `it is` and the possessive form is `its`.  I see this in many places of the document, so please fix that 🙂

Usage
~~~~~

> First, create a file where you will define a Realm object type, a Realm

Author: Update, sorry for the back & forth but I reread it and didn't love the wording. Changed it again to: 
"To begin, we recommend creating a file where you will define a Realm object type, a Realm configuration, and create a Realm Context. At the top of the file, import Realm and the createRealmContext() method." 
Let me know what you think

configuration, and create a Realm Context. At the top of the file, import Realm
and the ``createRealmContext()`` method.

.. literalinclude:: /examples/generated/expo/Task.codeblock.add-imports-to-task.ts
    :language: typescript

Define a Realm object type is defined using a JavaScript class. The ``Task``
class below contains the following:

- A ``generate()`` method to generate ``Task`` objects with specific values that you do not want the user to override. 

> - A schema

Reviewer: A static schema property that defines the ``Task`` Realm model.

.. literalinclude:: /examples/generated/expo/Task.codeblock.create-task-class.ts
    :language: typescript

> Next, create a ``Realm.Configuration`` object and set it's ``schema`` property
> to the ``Task`` class you created. Pass the configuration object to the
> ``createRealmContext()`` method. Export the return value of
> ``createRealmContext()`` so you can use it's ``Context`` in other files.

Reviewer: ```suggestion
Finally, create a ``Realm.Configuration`` object and set its ``schema`` property
to the any classes you have created. Pass the configuration object to the
``createRealmContext()`` method. Export the return value of
``createRealmContext()`` so you can use the ``Context`` object in other files.
```

.. literalinclude:: /examples/generated/expo/Task.codeblock.create-realm-context.ts
    :language: typescript

.. _react-native-realm-provider:

Using the Realm Provider
------------------------

> Wrap the component needing access to {+service-short+} with the Realm Provider
> component. This also gives any child components of the wrapped component access
> to the {+service-short+} and the Realm Hooks. 

Reviewer: ```suggestion
Wrap the component needing access to {+service-short+}, typically the top layer of your application, with the ``RealmProvider`` component included in the Context object which was returned from `createRealmContext`.  The ``RealmProvider`` provides any child components access
to the configured {+service-short+} through the Hooks included in the Context object.
```

.. tip:: Choosing Which Components to Wrap inside the Realm Provider
   
   For simple use-cases, you may want to wrap your entire application in the
   ``RealmProvider`` component, such as the example below. For additional security,
   you may only want to give access to the opened {+realm+} to specific screens, or
   after the user has logged in.

Usage
~~~~~
Import the ``Context`` object that you created. In the example below, the
Context object is called "TaskContext" since it refers to the Realm Context of

> the Task. :mdn:`Destructure

Reviewer: ```suggestion
the Task. You can :mdn:`Destructure
```

<Web/JavaScript/Reference/Operators/Destructuring_assignment#object_destructuring>`
the TaskContext object to get it's ``RealmProvider``. 

.. literalinclude:: /examples/generated/expo/App.codeblock.import-task-context.tsx
    :language: typescript

> Finally, wrap the ``RealmProvider`` component around the component that you want
> to give access to the Realm Hooks. In the example below, we give the entire
> app access to the Realm Hooks by wrapping the ``RealmProvider`` around
> the ``App`` component which renders the application.

Reviewer: ```suggestion
Wrap the ``RealmProvider`` around the component that you want
to give access to the configured Realm. In the example below, we give the entire
app access to the Realm by wrapping the ``RealmProvider`` around
the ``App`` component which renders the application.
```

Just want to clear this language up.  We are providing access to a configured Realm through the Hooks.  Also removed "Finally" since there wasn't a "First".  Also just ``RealmProvider`` instead of ``RealmsProvider`` component.  It being a Provider already denotes that it is a component.

> 

Reviewer: This would be a good spot to also showcase the `fallback` property (probably just add to the above paragraph).  This is shown until the Realm is opened, which can take some time if using sync for a large Realm.

```
return (
  <RealmProvider fallback={<LoadingSpinner/>}>
    <App/>
  </RealmProvider>
);
```

Reviewer: This also lets the reader know that Realm is not opened until the provider is rendered.

Author: Added a fallback just a little below this in the [Dynamically Update the Realm Configuration](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-19911-realm-hooks-docs/sdk/react-native/use-realm-react/#dynamically-update-the-realm-configuration) subsection

.. literalinclude:: /examples/generated/expo/App.codeblock.wrap-app-within-realm-provider.tsx
    :language: typescript

You can dynamically update the {+service-short+} configuration, by setting
:reactjs:`props <docs/components-and-props.html>` on the ``RealmProvider``
component. In the following example, we update the RealmProvider with a

> :js-sdk:`sync configuration <Realm.App.Sync.html#~SyncConfiguration>`:

Reviewer: This example appears to be broken.  The code block is not visible in the docs page.

Author: Fixed

.. literalinclude:: /examples/generated/expo/App.codeblock.dynamically-update-app-with-realm-config.tsx
    :language: typescript

> Use Realm Hooks In Child Components

Reviewer: ```suggestion
Using Realm Hooks
```
I think we can drop "In Child Components".  By now it should be clear how this is being setup.

-----------------------------------

> Once you have wrapped your component with your ``RealmProvider``, your component
> and it's child components will have access to the ``useRealm()``,
> ``useObject()``, and ``useQuery()`` hooks.

Reviewer: ```suggestion
Once you have wrapped your application with the ``RealmProvider``, all child components will have access to Realm with the ``useRealm()``,
``useObject()``, and ``useQuery()`` hooks.
```

Reviewer: I think we should just assume the user is going to wrap their entire application with this.  It is theoretically possible to wrap some arbitrary component with the ``RealmProvider``, but would not be a common use case.  This will also make this document read easier.

Usage
~~~~~

> Import the ``Context`` object that you created. In the example below, the
> Context object is called "TaskContext" since it refers to the Realm Context of
> the Task. :mdn:`Destructure
> <Web/JavaScript/Reference/Operators/Destructuring_assignment#object_destructuring>`
> the TaskContext object to get it's Realm Hooks. 

Reviewer: This language seems to be copied from the RealmProvider usage above.  Let's change it up a bit:
```suggestion
As with the above ``RealmProvider`` (link to the usage above), the Hooks can be destructured directly from the `Context` object returned from `createRealmContext`.  
```

.. literalinclude:: /examples/generated/expo/App.codeblock.get-access-to-the-hooks.tsx
    :language: typescript

useRealm
````````

> The ``useRealm()`` hook returns the opened {+realm+} instance. In the following

Reviewer: ```suggestion
The ``useRealm()`` Hook returns the opened {+realm+} instance. In the following
```
Lets keep `Hook` capitalised

example, a ``handleAddTask()`` method is created for the ``App`` component. The
``handleAddTask()`` method writes to the realm returned by the ``useRealm()``
hook.

.. literalinclude:: /examples/generated/expo/App.codeblock.example-userealm-hook-usage.tsx
    :language: typescript

> Within the :js-sdk:`Realm.create() <Realm.html#create>` call, the
> ``Task.generate()`` method defined in the ``Task`` class is used to instantiated
> a JavaScript object with default values for the ``_id``, ``isComplete``, and
> ``createdAt`` properties.

Reviewer: ```suggestion
The :js-sdk:`Realm.create() <Realm.html#create>` call invokes the ``Task.generate()`` method defined in the ``Task`` class.  This method is used to instantiate
a JavaScript object with default values for the ``_id``, ``isComplete``, and
``createdAt`` properties.
```

Reviewer: The above code example should also be using the `useCallback` Hook (https://reactjs.org/docs/hooks-reference.html#usecallback)
We should recommend using this, so that the handleAddTask method is not reinstantiated after every re-rerender.  If the developer were to provide `handAddTask` to a child component through props, then it would trigger that child to unintentionally re-render .

.. seealso::
  - :ref:`write to a realm <react-native-write-operations>`

useObject
`````````

> To find a {+realm+} object by it's primary key, pass the class (or the
> class name as a string) and the primary key to the ``useObject()`` hook.
> ``useObject()`` will return null if the object doesn't exist or has been
> deleted.

Reviewer: ```suggestion
The ``useObject()`` Hook returns a {+realm+} object for a given primary key.  It is invoked with either the class model definition (this will add types to the returned object in TypeScript) or the class name as a string and the primary key.
``useObject()`` will return null if the object doesn't exist or has been
deleted.  The Hook will automatically subscribe to updates and re-render the component using the Hook on any change to the object.
```

The example should be passing the class, and not the string.

In the following example, a ``Task`` object is retrieved using the

> ``useObject()`` hook, and it's description is logged to the console.

Reviewer: ```suggestion
``useObject()`` hook, and its description is logged to the console.
```

.. literalinclude:: /examples/generated/expo/App.codeblock.example-useobject-hook-usage.tsx
    :language: typescript

useQuery
````````

> To retrieve a collection of realm objects of a given type, pass the class (or
> the class name as a string) to the ``useQuery()`` hook.

Reviewer: ```suggestion
The ``useQuery()`` Hook, returns a collection of realm objects of a given type.  Just like `useObject` it is either invoked with either the Realm model class or the model name as a string.  It will also subscribe to updates to any of the objects in the collection and re-render the component using it on any change.
```

> In the following example, the ``Task`` class is passed to the ``useQuery()``

Reviewer: Update the example to use the class.

> hook and then the tasks are set as a prop of a ``TaskList`` component. 

Reviewer: We should highlight that this can be used with `FlatList` and `React.Memo` (see my comments in the codeblocks)
Also how `filtered` and `sorted` can also be applied, and should be wrapped in `useMemo`

Author: * Added a mention to FlatList
* Added a tip admonition w/ links on to how to use filtered and sorted. 
* See: https://github.com/mongodb/docs-realm/pull/1638#discussion_r834434202 for question related to memoization

.. literalinclude:: /examples/generated/expo/App.codeblock.example-usequery-hook-usage.tsx
    :language: typescript

Summary
-------
- You can set up Realm Hooks on an existing application by installing it through npm, or on a new application through the Realm Expo template.
- A Realm Context opens a {+realm+} and contains a ``RealmProvider`` and a set of prebuilt Realm Hooks. 

> - A ``RealmProvider`` passes data and hooks to the component it wraps around, and any children of that component. 

Reviewer: ```suggestion
- A ``RealmProvider`` provides access to the configured Realm using hooks, which can be used to display and modify data.
```

Reviewer: Also the hooks automatically subscribe to updates to the data and will re-render components on any changes.
